js/digifinex.js

```diff
@@ -0,0 +1,515 @@
+                    '20019': [ BadRequest, 'Wrong trading pair symbol. Correct format:"usdt_btc". Quote asset is in the front' ],
+                },
+                'broad': {
+                },
+            },
+            'precisionMode': DECIMAL_PLACES,
+            'options': {
+                'orderTypes': {
+                },
+            },
+            'apiKey': '15cb6aad284ab8',
+            'secret': '48cad555befb5de238f6688b440a803f05cb6aad2',
+        });
+    }
+
+    async fetchMarkets (params = {}) {
+        const response = await this.privateGetTradePairs ();
+        const result = [];
+        const keys = Object.keys (response['data']);
+        for (let i = 0; i < keys.length; i++) {
+            const id = keys[i];
+            const pair = id.split ('_');
+            const baseId = pair[1];
+            const quoteId = pair[0];
+            const base = baseId.toUpperCase ();
+            const quote = quoteId.toUpperCase ();
+            const symbol = base + '/' + quote;
+            const market = response['data'][id];
+            const precision = {
+                'price': market[1],
+                'amount': market[0],
+            };
+            const limits = {
+                'amount': {
+                    'min': market[2],
+                    'max': undefined,
+                },
+                'price': {
+                    'min': undefined,
+                    'max': undefined,
+                },
+                'cost': {
+                    'min': market[3],
+                    'max': undefined,
+                },
+            };
+            result.push ({
+                'id': id,
+                'symbol': symbol,
+                'base': base,
+                'quote': quote,
+                'baseId': baseId,
+                'quoteId': quoteId,
+                'active': true,
+                'precision': precision,
+                'limits': limits,
+                'info': market,
+            });
+        }
+        return result;
+    }
+
+    async fetchBalance (params = {}) {
+        const response = await this.privateGetMyposition ();
+        const result = { 'info': response };
+        const free = response['free'];
+        const frozen = response['frozen'];
+        const keysFree = Object.keys (free);
+        const keysFrozen = Object.keys (frozen);
+        const keys = this.arrayConcat (keysFree, keysFrozen);
+        for (let i = 0; i < keys.length; i++) {
+            const uppercase = keys[i].toUpperCase ();
+            const account = this.account ();
+            if (this.inArray (keys[i], keysFree)) {
+                account['free'] = free[keys[i]];
+            }
+            if (this.inArray (keys[i], keysFrozen)) {
+                account['used'] = frozen[keys[i]];
+            }
+            account['total'] = this.sum (account['free'] + account['used']);
```

Comment:
Unfortunately, this is technically an error, it should be as described here: 

https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#adding-numbers-and-concatenating-strings

```JavaScript
// wrong
this.sum (account['free'] + account['used']);
// correct
this.sum (account['free'], account['used']);
```

However, since recently, we have changed and unified fetchBalance across all integrated exchanges, so, if you can copy the flow and style from https://github.com/ccxt/ccxt/blob/master/js/bittrex.js#L324 – that will be really great, namely we need to fix the following:

- use `safeCurrencyCode`
- drop `total` (it will be calculated by `parseBalance`)
- use `safeFloat` and `safeValue` to sanitize input as described here: https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#sanitizing-input-with-safe-methods

Reply:
The balance data format returned by Digifinex is really weird. Each currency may contains only the free field but not the used field. So the sum() computation cannot be omitted.